### PR TITLE
Feature/arima

### PR DESCRIPTION
**Pull request by StanislasGuinel**
PR to add ARIMA model:
- one auto_arima model per timeseries
- external features support
- can automatically find 'm' (seasonality) using gluonts.time_feature.get_seasonality if 'm' not set by user

### CODE REVIEW

**Comment by StanislasGuinel** [comment]
Because I merged with feature/timestamp-preparation (sorry about that), we should wait for the feature/timestamp-preparation branch to be merged with release/1.0 before reviewing this one.

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (alexcombessie) [inline custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json:582]: I would recommend a description, since all other parameters have one.

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (alexcombessie) [inline custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json:598]: Personally, not a fan of making "timeseries" into a portmanteau word. Just a convention, but if we were to follow it, then our users should be called "datascientist" etc.
```suggestion
            "description": "Number of last records to extract per time series",
```

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (alexcombessie) [inline custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json:602]: 1 point would make life difficult for models, I would suggest a minimum of 4.

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [inline python-lib/custom_gluon_models/arima.py:106]: Why not let trace to True?

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [inline python-lib/custom_gluon_models/arima.py:68]: Why is `alpha` hardcoded? Shouldn't we pickup the settings chosen by the user in the Predict recipe?

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [inline python-lib/custom_gluon_models/arima.py:49]: Could the logger be a bit more specific?

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [inline python-lib/custom_gluon_models/arima.py:101]:
```suggestion
        logger.info("Training one Auto ARIMA model per time series ...")
```

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [inline python-lib/custom_gluon_models/arima.py:81]: To be exact, that model shouldn't be named Arima, but AutoARIMA. The reason is that Arima has a specific meaning in the literature. If you look at https://alkaline-ml.com/pmdarima/modules/classes.html there is one class for ARIMA and another for AutoARIMA.

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [inline python-lib/custom_gluon_models/arima.py:18]:
```suggestion
class AutoARIMAPredictor(RepresentablePredictor):
```

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (alexcombessie) [inline custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json:396]:
```suggestion
            "label": "- Auto ARIMA",
```

File: python-lib/dku_io_utils/recipe_config_loading.py
Comment (alexcombessie) [inline python-lib/dku_io_utils/recipe_config_loading.py:130]:
```suggestion
        if params["max_timeseries_length"] < 4:
```

File: python-lib/dku_io_utils/recipe_config_loading.py
Comment (alexcombessie) [inline python-lib/dku_io_utils/recipe_config_loading.py:131]:
```suggestion
            raise PluginParamValidationError("Number of records must be higher than 4")
```

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json:598]: ah good point ! i will make it into 2 words

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [reply python-lib/custom_gluon_models/arima.py:106]: Hum ok, but then, we should allow users to set trace=True in the kwargs. In the current setup, if a user tries it, the recipe will fail because of duplicate kwards.

File: python-lib/custom_gluon_models/arima.py
Comment (StanislasGuinel) [reply python-lib/custom_gluon_models/arima.py:68]: I hardcoded alpha to generate 99 samples that are all the percentiles from 1 to 99.
A GluonTS model outputs during prediction a SampleForecast class that contains many prediction samples generated by the model. These samples are then used to compute quantiles that I retrieve later based on the user input.

File: python-lib/custom_gluon_models/arima.py
Comment (StanislasGuinel) [reply python-lib/custom_gluon_models/arima.py:106]: yes i can do that :)

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [reply python-lib/custom_gluon_models/arima.py:68]: Okay! If this is to allow the computation of quantile-based performance metrics, that's fine. 

However, how can we ensure the correct alpha is computed when the user chooses a confidence interval in the predict recipe?

Comment (StanislasGuinel) [reply python-lib/custom_gluon_models/arima.py:68]: when a user chooses a confidence interval, I convert it into 2 quantiles (90% confidence interval becomes quantiles 0.05 and 0.95), then I call `sample_forecasts.quantile_ts(quantile)` for each of the 2 quantiles, where `sample_forecasts` is basically the output of `predictor.predict(gluon_dataset)`. So the 2 quantiles retrieved from  `sample_forecasts` are quantiles I set before with the alpha.

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (alexcombessie) [inline custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json:397]:
```suggestion
            "description": "Based on the pyramid-arima package. Slow for multiple time series.",
```

File: python-lib/custom_gluon_models/arima.py
Comment (alexcombessie) [reply python-lib/custom_gluon_models/arima.py:68]: Ok super, closing this then

File: python-lib/gluonts_forecasts/model_handler.py
Comment (alexcombessie) [inline python-lib/gluonts_forecasts/model_handler.py:97]:
```suggestion
    "autoarima": {
```

File: python-lib/gluonts_forecasts/model_handler.py
Comment (alexcombessie) [inline python-lib/gluonts_forecasts/model_handler.py:98]:
```suggestion
        LABEL: "AutoARIMA",
```

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json]: 545a70f

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json]: 545a70f

File: python-lib/custom_gluon_models/arima.py
Comment (StanislasGuinel) [reply python-lib/custom_gluon_models/arima.py]: 545a70f
trace=False by default actually, so i just removed it

File: python-lib/custom_gluon_models/arima.py
Comment (StanislasGuinel) [reply python-lib/custom_gluon_models/arima.py]: 545a70f
and the logger is initialised with "Auto ARIMA"

File: custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json]: 545a70f

**Comment by StanislasGuinel** [comment]
One comment about seasonality, I used the gluon method gluonts.time_feature.get_seasonality, but for Day and Week frequency, there is no seasonality (it is set to 1). Here is the values they use:
```
DEFAULT_SEASONALITIES = {
    "H": 24,
    "D": 1,
    "W-SUN": 1,
    "M": 12,
    "B": 5,
    "Q-DEC": 4,
}
```
Should we overwrite their method and set for example "D": 7 and "W": 52 or just keep it that way and let advanced users set "m" to the value they want (and we will explain this "m" value in the documenation) ?

**Comment by alexcombessie** [comment]
Since it's our implementation, I think what you suggest is a good idea! And in case the user sets the m value, we drop the default behavior.

File: python-lib/gluonts_forecasts/model_handler.py
Comment (alexbourret) [reply python-lib/gluonts_forecasts/model_handler.py]: Careful to change the key in recipe.json too if you change this one

**Review by alexbourret — APPROVED** [review]

**Review by alexcombessie — APPROVED** [review]
LGTM!

A very small suggestion before you merge: Since the UX name we chose is "AutoARIMA" I think we could harmonize our logging ("Auto ARIMA" -> "AutoARIMA")

**Review by alexcombessie — APPROVED** [review]

**Pull request merged**
